components/pd_client/src/client.rs

```diff
@@ -398,25 +402,23 @@ impl PdClient for RpcClient {
             tx.unbounded_send(req)
                 .unwrap_or_else(|e| panic!("send request to unbounded channel failed {:?}", e));
             inner.hb_sender = Either::Right(tx);
-            Box::new(
-                sender
+            Box::pin(async move {
+                let result = sender
+                    .sink_compat()
                     .sink_map_err(Error::Grpc)
-                    .send_all(rx.then(|r| match r {
-                        Ok(r) => Ok((r, WriteFlags::default())),
-                        Err(()) => Err(Error::Other(box_err!("failed to recv heartbeat"))),
-                    }))
-                    .then(|result| match result {
-                        Ok((mut sender, _)) => {
-                            info!("cancel region heartbeat sender");
-                            sender.get_mut().cancel();
-                            Ok(())
-                        }
-                        Err(e) => {
-                            error!("failed to send heartbeat"; "err" => ?e);
-                            Err(e)
-                        }
-                    }),
-            ) as PdFuture<_>
+                    .send_all(&mut rx.map(|r| Ok((r, WriteFlags::default()))))
```

Comment:
Will it still return early with nice message if rx gets error?

Reply:
The rx in futures 0.1 will not receive any errors, `type Error = ()`. The receiver will only have two states, `Some(item)` or `None`